### PR TITLE
Added combine publisher extensions to support combining 5 publishers

### PR DESCRIPTION
**Pull request by heecheon92**
## Checklist

- [x] I checked the [**Contributing Guidelines**](https://github.com/SwifterSwift/SwifterSwift/blob/master/CONTRIBUTING.md) before creating this request.
- [x] New extensions are written in Swift 5.0.
- [x] New extensions support iOS 10.0+ / tvOS 9.0+ / macOS 10.10+ / watchOS 2.0+, or use `@available` if not.
- [x] I have added tests for new extensions, and they passed.
- [x] All extensions have a **clear** comments explaining their functionality, all parameters and return type in English.
- [x] All extensions are declared as **public**.
- [x] I have added a [changelog](https://github.com/SwifterSwift/SwifterSwift/blob/master/CHANGELOG_GUIDELINES.md) entry describing my changes.

### CODE REVIEW

File: Sources/SwifterSwift/Combine/PublisherExtensions.swift [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    ///       Result: (8, 2, 9, 3, 1).
+    ///       Result: (8, 2, 21, 3, 1).
+    ///
+    /// - Parameters:
+    ///     - publisher1: A second publisher to combine with the first publisher.
+    ///     - publisher2: A third publisher to combine with the first publisher.
+    ///     - publisher3: A fourth publisher to combine with the first publisher.
+    ///     - publisher4: A fifth publisher to combine with the first publisher.
+    /// - Returns: A publisher that receives and combines elements from this publisher and four other publishers.
+    func combineLatest<P, Q, R, Y>(
```
Comment (LucianoPAlmeida) [inline Sources/SwifterSwift/Combine/PublisherExtensions.swift:49]: Why 4 specifically? I imagine that combine already has the ones 2 and 3 right?
Using [parameter packs](https://github.com/apple/swift-evolution/blob/main/proposals/0393-parameter-packs.md) wouldn't be a more generic way to handle this?

Comment (heecheon92) [reply Sources/SwifterSwift/Combine/PublisherExtensions.swift:49]: Yes, Combine already provides "combineLatest" for 2 to 3 additional publishers and having additional 4th publisher comes in handy time to time.

I have tried to make reimplementation of combineLatest with parameter packs as you suggested but the feature is not mature enough to satisfy protocol constraint.

스크린샷 2023-09-29 오전 9 10 48

For example, each publisher in combineLatest must satisfy that the associated Failure must be identical to each iterating generic placeholder but currently Swift doesn't provide a way to put that constraint.

**Review by LucianoPAlmeida — APPROVED** [review]
Looks good,
Can you fix the conflicts?

**Comment by heecheon92** [comment]
@LucianoPAlmeida 

I just fixed the sync issue.

**Review by LucianoPAlmeida — DISMISSED** [review]

**Comment by heecheon92** [comment]
I see that SwifterSwift/Darwin fails with the following error.

```
bundler: failed to load command: danger (/usr/local/lib/ruby/gems/3.0.0/bin/danger)
/usr/local/lib/ruby/gems/3.0.0/gems/octokit-4.25.1/lib/octokit/response/raise_error.rb:14:in `on_complete': POST https://api.github.com/repos/SwifterSwift/SwifterSwift/issues/1124/comments: 500 -  (Octokit::InternalServerError)
	from /usr/local/lib/ruby/gems/3.0.0/gems/faraday-1.10.3/lib/faraday/middleware.rb:19:in `block in call'
	from /usr/local/lib/ruby/gems/3.0.0/gems/faraday-1.10.3/lib/faraday/response.rb:61:in `on_complete'
```

Is this something that needs to be worked on my end?

**Review by LucianoPAlmeida — APPROVED** [review]

File: Sources/SwifterSwift/Combine/PublisherExtensions.swift [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+//  PublisherExtensions.swift
+//  SwifterSwift
+//
+//  Created by Heecheon Park on 7/28/23.
+//  Copyright © 2023 SwifterSwift
+//
+
+// swiftlint:disable large_tuple
+#if canImport(Foundation) && canImport(Combine)
+import Foundation
```
Comment (guykogus) [inline Sources/SwifterSwift/Combine/PublisherExtensions.swift:11]: Do you need `Foundation`?
```suggestion
#if canImport(Combine)
```

File: Tests/CombineTests/PublisherExtensionsTests.swift [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+//
+
+// swiftlint:disable identifier_name
+// swiftlint:disable large_tuple
+
+@testable import SwifterSwift
+import XCTest
+
+#if canImport(Foundation) && canImport(Combine)
+import Foundation
```
Comment (guykogus) [inline Tests/CombineTests/PublisherExtensionsTests.swift:16]: Same comment
```suggestion
#if canImport(Combine)
```

File: Tests/CombineTests/PublisherExtensionsTests.swift [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+            XCTAssert(result[3] == (2, 16, nil, false, "ZZ"))
+            XCTAssert(result[4] == (2, 16, 1, false, "ZZ"))
+            XCTAssert(result[5] == (2, 16, 1, true, "ZZ"))
+        }
+    }
+}
+#endif
+
+// swiftlint:enable large_tuple
+// swiftlint:enable identifier_name
```
Comment (guykogus) [inline Tests/CombineTests/PublisherExtensionsTests.swift:126]:
```suggestion
// swiftlint:enable large_tuple identifier_name
```

File: Tests/CombineTests/PublisherExtensionsTests.swift [hunk 5]
```diff
@@ -0,0 +1,126 @@
+//
+//  PublisherExtensionsTests.swift
+//  SwifterSwift
+//
+//  Created by Heecheon Park on 7/28/23.
+//  Copyright © 2023 SwifterSwift
+//
+
+// swiftlint:disable identifier_name
+// swiftlint:disable large_tuple
```
Comment (guykogus) [inline Tests/CombineTests/PublisherExtensionsTests.swift:10]:
```suggestion
// swiftlint:disable identifier_name large_tuple
```

File: Sources/SwifterSwift/Combine/PublisherExtensions.swift [hunk 6]
```diff
@@ -0,0 +1,123 @@
+//
+//  PublisherExtensions.swift
+//  SwifterSwift
+//
+//  Created by Heecheon Park on 7/28/23.
+//  Copyright © 2023 SwifterSwift
+//
+
+// swiftlint:disable large_tuple
```
Comment (guykogus) [inline Sources/SwifterSwift/Combine/PublisherExtensions.swift:9]: Don't disable for the whole file, only the function where you need it.

File: Tests/CombineTests/PublisherExtensionsTests.swift (hunk 5 above)
Comment (guykogus) [reply Tests/CombineTests/PublisherExtensionsTests.swift:10]: Also, disable only for the functions that you need it, not the whole file.

**Review by guykogus — CHANGES_REQUESTED** [review]
Looks great, a few small comments.
Also, we probably won't need this once we have parameter packs in Swift 5.9. Or at least we can have unlimited sizes.

**Comment by guykogus** [comment]
@heecheon92 actually can you please also update `README.md` and `README_CN.md` to include the new `Combine` section. You can basically copy from https://github.com/SwifterSwift/SwifterSwift/pull/1139/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5

File: Tests/CombineTests/PublisherExtensionsTests.swift (hunk 5 above)
Comment (heecheon92) [reply Tests/CombineTests/PublisherExtensionsTests.swift:10]: tbh, it was my first time using SwiftLint but I guess I get your point. Thanks,

File: Sources/SwifterSwift/Combine/PublisherExtensions.swift (hunk 6 above)
Comment (guykogus) [reply Sources/SwifterSwift/Combine/PublisherExtensions.swift:9]: @heecheon92 just move this down into extension , let's say line 15, and the enabling up to 120.

File: Tests/CombineTests/PublisherExtensionsTests.swift [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+@testable import SwifterSwift
+import XCTest
+
+#if canImport(Combine)
+import Combine
+
+final class PublisherExtensionsTests: XCTestCase {
+
+    func testCombineLatest() throws {
+        if #available(macOS 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *) {
```
Comment (guykogus) [inline Tests/CombineTests/PublisherExtensionsTests.swift:20]: Use `@available` for the class, same as you do in the implementation.

File: Tests/CombineTests/PublisherExtensionsTests.swift [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+            let p1 = PassthroughSubject<Int, Never>()
+            let p2 = PassthroughSubject<Int, Never>()
+            let p3 = PassthroughSubject<Int, Never>()
+            let p4 = PassthroughSubject<Int, Never>()
+            let p5 = PassthroughSubject<Int, Never>()
+            
+            let cancellable = p1.combineLatest(p2, p3, p4, p5)
+                .sink { result.append(($0, $1, $2, $3, $4)) }
+            
+            defer { cancellable.cancel() }
```
Comment (guykogus) [inline Tests/CombineTests/PublisherExtensionsTests.swift:32]: I don't think you need to do this. The `Cancellable` should do it automatically when it's deinitialised.

File: Tests/CombineTests/PublisherExtensionsTests.swift [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+            let p4 = PassthroughSubject<Bool, Never>()
+            let p5 = PassthroughSubject<Character, Never>()
+            
+            let cancellable = p1.combineLatest(p2, p3, p4, p5, { v1, v2, v3, v4, v5 in
+                return (v1, v2 * v2, v3 == nil ? nil : 1, v4, "\(v5)\(v5)")
+            })
+            .map { ($0 * 2, $1, $2, $3, $4 ) }
+            .sink { result.append(($0, $1, $2, $3, $4)) }
+            
+            defer { cancellable.cancel() }
```
Comment (guykogus) [inline Tests/CombineTests/PublisherExtensionsTests.swift:87]: Same comment here, shouldn't be necessary.

File: Tests/CombineTests/PublisherExtensionsTests.swift [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+            var result: [(Int, Int, Int?, Bool, String)] = []
+
+            let p1 = PassthroughSubject<Int, Never>()
+            let p2 = PassthroughSubject<Int, Never>()
+            let p3 = PassthroughSubject<Int?, Never>()
+            let p4 = PassthroughSubject<Bool, Never>()
+            let p5 = PassthroughSubject<Character, Never>()
+            
+            let cancellable = p1.combineLatest(p2, p3, p4, p5, { v1, v2, v3, v4, v5 in
+                return (v1, v2 * v2, v3 == nil ? nil : 1, v4, "\(v5)\(v5)")
```
Comment (guykogus) [inline Tests/CombineTests/PublisherExtensionsTests.swift:82]:
```suggestion
                (v1, v2 * v2, v3 == nil ? nil : 1, v4, "\(v5)\(v5)")
```

File: Tests/CombineTests/PublisherExtensionsTests.swift [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+            XCTAssert(result.count == 6)
+            XCTAssert(result[4] == (11, 22, 33, 4, 55))
+            XCTAssert(result[5] == (11, 22, 33, 44, 55))
+        }
+    }
+    
+    func testCombineLatestTransform() throws {
+        if #available(macOS 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *) {
+            
+            var result: [(Int, Int, Int?, Bool, String)] = []
```
Comment (guykogus) [inline Tests/CombineTests/PublisherExtensionsTests.swift:73]:
```suggestion
            var result = [(Int, Int, Int?, Bool, String)]()
```

File: Tests/CombineTests/PublisherExtensionsTests.swift [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+import XCTest
+
+#if canImport(Combine)
+import Combine
+
+final class PublisherExtensionsTests: XCTestCase {
+
+    func testCombineLatest() throws {
+        if #available(macOS 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *) {
+            var result: [(Int, Int, Int, Int, Int)] = []
```
Comment (guykogus) [inline Tests/CombineTests/PublisherExtensionsTests.swift:21]:
```suggestion
            var result = [(Int, Int, Int, Int, Int)]()
```

File: Tests/CombineTests/PublisherExtensionsTests.swift [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+                .sink { result.append(($0, $1, $2, $3, $4)) }
+            
+            defer { cancellable.cancel() }
+            
+            p1.send(1)
+            p2.send(2)
+            p3.send(3)
+            p4.send(4)
+            
+            XCTAssertTrue(result.isEmpty)
```
Comment (guykogus) [inline Tests/CombineTests/PublisherExtensionsTests.swift:39]:
```suggestion
            XCTAssert(result.isEmpty)
```

File: Tests/CombineTests/PublisherExtensionsTests.swift [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+                return (v1, v2 * v2, v3 == nil ? nil : 1, v4, "\(v5)\(v5)")
+            })
+            .map { ($0 * 2, $1, $2, $3, $4 ) }
+            .sink { result.append(($0, $1, $2, $3, $4)) }
+            
+            defer { cancellable.cancel() }
+            
+            p1.send(1)
+            p2.send(2)
+            XCTAssertTrue(result.isEmpty)
```
Comment (guykogus) [inline Tests/CombineTests/PublisherExtensionsTests.swift:91]:
```suggestion
            XCTAssert(result.isEmpty)
```

File: Tests/CombineTests/PublisherExtensionsTests.swift [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+            p4.send(false)
+            p5.send("A")
+            XCTAssertFalse(result.isEmpty)
+            XCTAssert(result.allSatisfy({
+                $0.0 == 2 &&
+                $0.1 == 4 &&
+                $0.2 == nil &&
+                $0.3 == false &&
+                $0.4 == "AA"
+            }))
```
Comment (guykogus) [inline Tests/CombineTests/PublisherExtensionsTests.swift:103]: Does this compile?
```suggestion
            XCTAssert(result.allSatisfy {
                $0.0 == 2 &&
                $0.1 == 4 &&
                $0.2 == nil &&
                $0.3 == false &&
                $0.4 == "AA"
            })
```

**Review by guykogus — CHANGES_REQUESTED** [review]

File: Sources/SwifterSwift/Combine/PublisherExtensions.swift (hunk 6 above)
Comment (heecheon92) [reply Sources/SwifterSwift/Combine/PublisherExtensions.swift]: fixed at bf46972a2f85cb9ffc39518a74154b24f76c5793

File: Tests/CombineTests/PublisherExtensionsTests.swift (hunk 7 above)
Comment (heecheon92) [reply Tests/CombineTests/PublisherExtensionsTests.swift:21]: handled at c66c7fb4bb52f15ec1c5b34be6c90ef77e380ad5

File: Tests/CombineTests/PublisherExtensionsTests.swift (hunk 8 above)
Comment (heecheon92) [reply Tests/CombineTests/PublisherExtensionsTests.swift]: handled at c66c7fb4bb52f15ec1c5b34be6c90ef77e380ad5

File: Tests/CombineTests/PublisherExtensionsTests.swift (hunk 9 above)
Comment (heecheon92) [reply Tests/CombineTests/PublisherExtensionsTests.swift]: handled at c66c7fb4bb52f15ec1c5b34be6c90ef77e380ad5

**Review by LucianoPAlmeida — APPROVED** [review]

File: Tests/CombineTests/PublisherExtensionsTests.swift (hunk 7 above)
Comment (guykogus) [reply Tests/CombineTests/PublisherExtensionsTests.swift:21]: You can remove the `if #available` now.

File: Tests/CombineTests/PublisherExtensionsTests.swift [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+            p3.send(33)
+            p4.send(44)
+            XCTAssert(result.count == 6)
+            XCTAssert(result[4] == (11, 22, 33, 4, 55))
+            XCTAssert(result[5] == (11, 22, 33, 44, 55))
+        }
+    }
+    
+    func testCombineLatestTransform() throws {
+        if #available(macOS 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *) {
```
Comment (guykogus) [inline Tests/CombineTests/PublisherExtensionsTests.swift:70]: Same here, remove it.

File: Tests/CombineTests/PublisherExtensionsTests.swift [hunk 17]
```diff
@@ -0,0 +1,120 @@
+//
+//  PublisherExtensionsTests.swift
+//  SwifterSwift
+//
+//  Created by Heecheon Park on 7/28/23.
+//  Copyright © 2023 SwifterSwift
+//
+
+// swiftlint:disable identifier_name
```
Comment (guykogus) [inline Tests/CombineTests/PublisherExtensionsTests.swift:9]: I don't love this covering the whole file. Maybe just use it for your tests.

**Review by guykogus — CHANGES_REQUESTED** [review]

**Comment by guykogus** [comment]
@heecheon92 closing due to inactivity, but feel free to reopen and finish the last comments.

**Pull request closed without merging**